packages/app/src/cli/services/generate.ts

```diff
@@ -150,4 +144,55 @@ function formatSuccessfulRunMessage(
   return options
 }
 
+async function handleTypeParameter(
+  type: string | undefined,
+  app: AppInterface,
+  templateSpecifications: TemplateSpecification[],
+): Promise<TemplateSpecification | undefined> {
+  if (!type) return
+
+  const templateSpecification = findTemplateSpecification(type, templateSpecifications)
+
+  if (!templateSpecification) {
+    const isShopifolk = await isShopify()
+    const allExternalTypes = getTypesExternalIdentitifier(templateSpecifications)
+    const tryMsg = isShopifolk ? 'You might need to enable some beta flags on your Organization or App' : undefined
+    throw new AbortError(
+      `Unknown extension type: ${type}.\nThe following extension types are supported: ${allExternalTypes.join(', ')}`,
+      tryMsg,
+    )
+  }
+
+  // Validate limits for selected type.
+  // If no type is selected, filter out any types that have reached their limit
+  templateSpecification.types.forEach((spec) => {
+    const existing = app.extensionsForType(spec)
+    const limit = spec.registrationLimit
+    if (existing.length >= limit) {
+      throw new AbortError(
+        'Invalid extension type',
+        `You can only generate ${limit} extension(s) of type ${spec.externalIdentifier} per app`,
+      )
+    }
+  })
+
+  return templateSpecification
+}
+
+async function getTemplateSpecifications(options: GenerateOptions): Promise<TemplateSpecification[]> {
+  const token = await ensureAuthenticatedPartners()
+  const apiKey = await ensureGenerateContext({...options, token})
+  const specifications = await fetchSpecifications({token, apiKey, config: options.config})
+  let templateSpecifications = convertSpecificationsToTemplate(specifications)
+  const remoteTemplateSpecifications = await fetchTemplateSpecifications(token)
+  templateSpecifications = templateSpecifications.concat(remoteTemplateSpecifications)
+  // If for any reason the remote template specifications API with the functions is not available, we will load the
+  // local functions specifications
+  if (remoteTemplateSpecifications.length === 0) {
+    const functions = await loadFunctionSpecifications(options.config)
+    templateSpecifications = templateSpecifications.concat(convertSpecificationsToTemplate(functions))
+  }
```

Comment:
We shouldn't need to do this, the API will be available and we can't keep duplicated hardcoded info in the CLI, the idea of this project was to remove that from the CLI :)

Reply:
https://github.com/Shopify/cli/pull/1541#discussion_r1140142058